ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/state/ValidatorIndexCache.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.datastructures.state;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import tech.pegasys.teku.bls.BLSPublicKey;
+import tech.pegasys.teku.infrastructure.collections.cache.Cache;
+import tech.pegasys.teku.infrastructure.collections.cache.LRUCache;
+import tech.pegasys.teku.infrastructure.collections.cache.NoOpCache;
+import tech.pegasys.teku.ssz.SSZTypes.SSZList;
+
+public class ValidatorIndexCache {
+  private final Cache<BLSPublicKey, Integer> validatorIndexes;
+  private final AtomicInteger lastIndex;
+
+  private static final int INDEX_NONE = -1;
+  static final ValidatorIndexCache NO_OP_INSTANCE =
+      new ValidatorIndexCache(NoOpCache.getNoOpCache(), INDEX_NONE);
+
+  @VisibleForTesting
+  ValidatorIndexCache(final Cache<BLSPublicKey, Integer> validatorIndexes, final int lastIndex) {
+    this.validatorIndexes = validatorIndexes;
+    this.lastIndex = new AtomicInteger(lastIndex);
+  }
+
+  public ValidatorIndexCache() {
+    this.validatorIndexes = new LRUCache<>(Integer.MAX_VALUE - 1);
+    this.lastIndex = new AtomicInteger(INDEX_NONE);
+  }
+
+  public Optional<Integer> getValidatorIndex(
+      final BeaconState state, final BLSPublicKey publicKey) {
+    final Optional<Integer> validatorIndex = validatorIndexes.getCached(publicKey);
+    if (validatorIndex.isPresent()) {
+      return validatorIndex.filter(index -> index < state.getValidators().size());
+    }
+
+    return findIndexFromState(state.getValidators(), publicKey);
+  }
+
+  private Optional<Integer> findIndexFromState(
+      final SSZList<Validator> validatorList, final BLSPublicKey publicKey) {
+    for (int i = Math.max(lastIndex.get(), 0); i < validatorList.size(); i++) {
```

Comment:
This isn't thread safe - it's possible that another thread updated the cache and lastIndex between when we checked the cache and getting the lastIndex here which could lead us to return unknown incorrectly.

We need to do `lastIndex.get()` and store the result, before we check the cache.  Then we are guaranteed that the cache is populated for all indices prior to the value we got.  We might have extra cache entries but that's safe.